aptos-move/framework/aptos-framework/sources/configs/Stake.move

```diff
@@ -0,0 +1,247 @@
+    }
+    /// Initialize the ValidatorInfo for account.
+    fun register_validator_candidate(account: &signer, consensus_pubkey: vector<u8>, network_address: vector<u8>) {
+        let stake_info = StakeInfo {
+            current_stake: 0,
+            active: Vector::empty(),
+            warmup: Vector::empty(),
+            cooldown: Vector::empty(),
+            inactive: Vector::empty(),
+        };
+        move_to(account, ValidatorInfo {
+            stake_info,
+            consensus_pubkey,
+            network_address,
+        });
+    }
+
+    /// Rotate the consensus key of the validator.
+    fun rotate_consensus_key(account: &signer, consensus_pubkey: vector<u8>) acquires ValidatorInfo, ValidatorSet {
+        let addr = Signer::address_of(account);
+        let validator_set = borrow_global_mut<ValidatorSet>(@CoreResources);
+        if (Vector::contains(&validator_set.validators, &addr)) {
+            on_new_epoch(validator_set);
+        };
+        borrow_global_mut<ValidatorInfo>(addr).consensus_pubkey = consensus_pubkey;
+    }
+
+    /// Initialize validator set to the core resource account.
+    fun initialize_validator_set(account: &signer, minimum_stake: u64, maximum_stake: u64) {
+        SystemAddresses::assert_core_resource(account);
+        move_to(account, ValidatorSet {
+            consensus_scheme: 0,
+            minimum_stake,
+            maximum_stake,
+            validators: Vector::empty(),
+            last_update_time_secs: Timestamp::now_seconds(),
+        });
+    }
+
+    /// Initiate by the validator info owner
+    fun add_validator(account: &signer) acquires ValidatorInfo, ValidatorSet {
+        let addr = Signer::address_of(account);
+        let validator_info = borrow_global_mut<ValidatorInfo>(addr);
+        let validator_set = borrow_global_mut<ValidatorSet>(@CoreResources);
+        assert!(!Vector::contains(&validator_set.validators, &addr), 0);
+
+        assert!(validator_info.stake_info.current_stake >= validator_set.minimum_stake, 0);
+        assert!(validator_info.stake_info.current_stake <= validator_set.maximum_stake, 0);
+
+        on_new_epoch(validator_set);
+        Vector::push_back(&mut validator_set.validators, addr);
+    }
+
+    /// Initiate by the validator info owner.
+    fun remove_validator(account: &signer) acquires ValidatorInfo, ValidatorSet {
+        let addr = Signer::address_of(account);
+        let validator_set = borrow_global_mut<ValidatorSet>(@CoreResources);
+
+        let (exist, index) = Vector::index_of(&validator_set.validators, &addr);
+        assert!(exist, 0);
+        on_new_epoch(validator_set);
+        Vector::swap_remove(&mut validator_set.validators, index);
+        assert!(Vector::length(&validator_set.validators) > 0, 0);
+    }
+
+    /// Triggers when validator set changes or after certain time period (in block prologue).
+    fun on_new_epoch(validator_set: &mut ValidatorSet) acquires ValidatorInfo {
```

Comment:
ideally epochs should be regularly paced over a period of time and not controllable by individual behavior

Reply:
I remember it's discussed 2 years ago, with regular pace we need to buffer all the changes and make them effective at the epoch.
it requires update to the new_epoch code every time instead of just calling reconfigure() with right permission.